### DISCUSSION TITLE
Installing in a Ubuntu docker image

### DISCUSSION THREAD

**Discussion by vulkanosaure** [uNetworking/uWebSockets.js#984] [Q&A]
Hi,

I'm trying to install uWebSockets.js on a vanilla Ubuntu docker image, in which I installed node
When running the container, I get the error

```
_Error: Cannot find module 'uWebSockets.js'

Require stack:

- /var/www/socket/main.js

    at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)

    at Module._load (node:internal/modules/cjs/loader:922:27)

    at Module.require (node:internal/modules/cjs/loader:1143:19)

    at require (node:internal/modules/cjs/helpers:119:18)

    at Object.<anonymous> (/var/www/socket/main.js:1:1)

    at Module._compile (node:internal/modules/cjs/loader:1256:14)

    at Module._extensions..js (node:internal/modules/cjs/loader:1310:10)

    at Module.load (node:internal/modules/cjs/loader:1119:32)

    at Module._load (node:internal/modules/cjs/loader:960:12)

    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:86:12) {

  code: 'MODULE_NOT_FOUND',

  requireStack: [ '/var/www/socket/main.js' ]

}_


```

Here's my dockerfile

```
FROM ubuntu

WORKDIR /var/www/socket

#install node
RUN apt-get update
RUN apt-get install -y ca-certificates curl gnupg
RUN mkdir -p /etc/apt/keyrings
RUN curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg
RUN echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_18.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list
RUN apt-get update
RUN apt-get install nodejs -y

#install curl + bun
RUN apt upgrade
RUN apt install curl
RUN apt install unzip
RUN curl -fsSL https://bun.sh/install | bash

#install uWebSockets
RUN ~/.bun/bin/bun install uNetworking/uWebSockets.js#v20.33.0

CMD ["node","main.js"]
```

During the creation process, i can see the message
` installed uWebSockets.js@github:uNetworking/uWebSockets.js#c10b47c`
Which seems to say that uWebSockets has been correctly installed

I'd went to the `node_modules` folder to confirm if the `uWebSockets.js` folder is there, 
The command `which node` returns `/usr/bin/node`
and `cd /usr/bin/node` returns `can't cd to /usr/bin/node`
as if it's a problem of access right

The content of `main.js` is the same as the example in the `readme`.

Any idea of how to troubleshoot this ?
Thanks in advance.

**Comment by vulkanosaure** [comment]
anyone ?

apologize if my question is naive but I tried my best

**Comment by vulkanosaure** [comment]
Ok i made it work,
i write down my findings if anyone come accross the same need.

to install and run from a Ubuntu docker image, here is my working docker file
```

FROM ubuntu

WORKDIR /var/www/socket

# install node 20

RUN apt update && apt upgrade
RUN apt install -y curl
RUN curl -fsSL https://deb.nodesource.com/setup_20.x | bash -
RUN apt install -y nodejs

# install curl + bun
RUN apt update && apt upgrade
RUN apt upgrade && apt install curl && apt install unzip
RUN curl -fsSL https://bun.sh/install | bash
RUN ~/.bun/bin/bun install uNetworking/uWebSockets.js#v20.33.0

CMD ["node","main.js"]

```

It's not ideal because there's a 60 second wait during the deb.nodesource.com setup that i didn't take the time to solve, but it works.
(Maybe starting from a node official docker image would be better)

and main.js should contain the example in the readme, 
but you must add :
`const { DEDICATED_COMPRESSOR_3KB } = require('uWebSockets.js');`
at the beginning.

**Comment by aditya-kumarr** [comment]
Still facing this issue, no matter what I try I get the "Cannot find module 'uWebsockets.js' or its corresponding type declarations."
And when I check docker files all dependencies are there but uWebsockets.js is empty.
Tried with bun, pnpm, yarn, npm, same results. It works fine in my mac but facing issue with node:20, node:20-slim and even alpine, There's barely any information on why this won't work.

**Reply by uNetworkingAB** [reply]
Docker is completely irrelevant to uWS. There is no difference, whatsoever, whether you install on Ubuntu in Docker or Ubuntu elsewhere.

You either have npm or you don't. If you have npm, then you install like in the README.md. That's it. Docker or not, irrelevant.

**Reply by aditya-kumarr** [reply]
I'm able to install every other package except for uWS, it works in my mac but it won't work in docker, the node_modules in docker contains every other dependency but the one for uWs is empty, 
I saw some threads saying it could be a docker issue because, the container needs to get correct binaries depending on the OS and if essential build tools are missing it could lead to this.

as I said earlier I've used every package manager it just won't work for node images.

This is my pacagke.json dependencies
 `"drizzle-orm": "^0.36.0",
   "jsonwebtoken": "^9.0.2",
   "uWebSockets.js": "github:uNetworking/uWebSockets.js#v20.51.0"
`
    
Edit: I tried a minimal setup, without my monorepo, the package is installed this time, but it seems like typescript is unable to pick up index.d.ts file and so cannot find any type declarations leading to that issue I faced.
I just checked all files are there but it's not picking up 

This is very weird issue because the same tsconfig file is working fine when trying to build from my machine.


Edit2: Nevermind I"m an idiot, I was importing from "uWebsockets.js" instead of "uWebSockets.js", had no idea macOs is case-insensitive by default, Anyways getting another error for related to Version, but atleast got this fixed,
Thanks for you help man. I was really confused an aimless before realising this 

> No. Install just takes the files in that release and puts in a folder. It's just a cp, in essence.

**Reply by uNetworkingAB** [reply]
> needs to get correct binaries depending on the OS and if essential build tools are missing it could lead to this.

No. Install just takes the files in that release and puts in a folder. It's just a cp, in essence.

**Comment by sitexw** [comment]
Just dropping this here, since it’s a problem I encountered not too long ago.  
I was getting this error while compiling my TypeScript code on a GitLab CI instance:  
```bash
Cannot find module 'uwebsockets.js' or its corresponding type declarations.
```  
Turns out I work on Windows, and locally (even with WSL), I had no compilation issues...  
It took me a little while to realize it was simply a case sensitivity issue.  
Since Windows is not case-sensitive, it had no trouble resolving the package path — unlike Linux.  
So to fix the problem, you just need to write the package name with the correct casing. For example:  
```typescript
import { App } from 'uWebSockets.js';
```
